src/test/java/com/amazon/ion/util/ZstdStreamInterceptorTest.java

```diff
@@ -0,0 +1,436 @@
+        return error;
+    }
+
+    /**
+     * Fails the test if the given reference points to any exception.
+     * @param error an error reference.
+     */
+    private void failOnAnyError(AtomicReference<Throwable> error) {
+        Throwable t = error.get();
+        if (t != null) {
+            Assertions.fail(t);
+        }
+    }
+
+    // This is a multithreaded test; execute multiple times to increase the chances of triggering a race condition if
+    // one exists.
+    @RepeatedTest(100)
+    public void differentClassLoadersInEachThread() throws Exception {
+        Thread withDefaultClassLoader = new Thread(() -> {
+            IonReaderBuilder readerBuilder = IonReaderBuilder.standard();
+            List<InputStreamInterceptor> streamInterceptors = readerBuilder.getInputStreamInterceptors();
+            assertEquals(1, streamInterceptors.size());
+            assertSame(GzipStreamInterceptor.INSTANCE, streamInterceptors.get(0));
+        });
+        AtomicReference<Throwable> withDefaultClassLoaderError = registerExceptionHandler(withDefaultClassLoader);
+        Thread withCustomClassLoader = new Thread(() -> {
+            AtomicReference<List<?>> streamInterceptors = new AtomicReference<>();
+            executeWithCustomClassLoader(
+                Thread.currentThread().getContextClassLoader(), // This will be our custom ClassLoader (set below)
+                (builderClass, builder) -> {
+                    streamInterceptors.set(assertGzipThen(builderClass, builder, ZstdStreamInterceptor.class));
+                }
+            );
+            // Verify that a new IonReaderBuilder instance does not re-detect the interceptors applicable to this
+            // ClassLoader.
+            executeWithCustomClassLoader(
+                Thread.currentThread().getContextClassLoader(), // This will be our custom ClassLoader (set below)
+                (builderClass, builder) -> {
+                    assertSame(
+                        streamInterceptors.get(),
+                        assertGzipThen(builderClass, builder, ZstdStreamInterceptor.class)
+                    );
+                }
+            );
+        });
+        AtomicReference<Throwable> withCustomClassLoaderError = registerExceptionHandler(withCustomClassLoader);
+        withCustomClassLoader.setContextClassLoader(new CustomInterceptorClassLoader());
+
+        withDefaultClassLoader.start();
+        withCustomClassLoader.start();
+
+        // While the spawned threads are working, verify they do not affect the parent thread.
+        IonReaderBuilder builder = IonReaderBuilder.standard().addInputStreamInterceptor(new LengthTooLongInterceptor(0));
+        assertGzipThen(builder, LengthTooLongInterceptor.class);
+
+        withDefaultClassLoader.join();
+        withCustomClassLoader.join();
+
+        failOnAnyError(withDefaultClassLoaderError);
+        failOnAnyError(withCustomClassLoaderError);
```

Comment:
Could this instead be phrased as a retrieval operation, fetching the list of interceptors from each thread? Then you could do the assertions in the main thread and each thread would be doing the same work (retrieving a list) absent the classloader configuration part.